sark/tseries.py

```diff
@@ -0,0 +1,171 @@
+from pathlib import Path
+from typing import List, TypeVar, Union
+from warnings import warn
+
+import pandas as pd
+
+_time_units = (
+    "days",
+    "day",
+    "hours",
+    "hour",
+    "minutes",
+    "minute",
+    "seconds",
+    "second",
+)
+
+_path_t = Union[str, Path]
+_col_t = TypeVar("_col_t", int, str)
+
+
+def read_timeseries(
+    fpath: _path_t,
+    *,
+    date_cols: Union[List[_col_t], None] = None,
+    col_units: Union[str, None] = None,
+    zero_idx: bool = False,
+    source_t: str = "",
+    **kwargs,
+):
+    """Read a time series from a file.
+
+    While the natural way to structure a time series dataset is with the index
+    column as datetime values, with subsequent columns holding other values,
+    there are a few other frequently used structures.
+
+    The first is to structure it as a table:
+
+    ===========  ===  ===  ====  ====  ====
+                  1    2    ..    23    24
+    ===========  ===  ===  ====  ====  ====
+     1/1/2016      0   10    ..   2.3   5.1
+     4/1/2016      3   11    ..   4.3   9.1
+    ===========  ===  ===  ====  ====  ====
+
+    When `source_t` is set to "table", this function reads a tabular dataset,
+    like the one above, and flattens it into a series, while setting the
+    appropriate datetime values as their index.
+
+    The other common structure is to split the datetime values into multiple
+    columns in the table:
+
+    ===========  ======  ======  ======
+      date        time    col1    col2
+    ===========  ======  ======  ======
+     1/1/2016     10:00    42.0    foo
+     4/1/2016     11:00    3.14    bar
+    ===========  ======  ======  ======
+
+    When `source_t` is set to "multicol". this function the table, while
+    combining the designated columns to construct the datetime values, which
+    are then set as the index.
+
+    If `source_t` is not specified (or set to an empty string), options
+    specific to this function are ignored, and all other keyword options are
+    passed on to the backend transparently; it's a noop.
+
+    Parameters
+    ----------
+    fpath : Union[str, Path]
+        Path to the dataset file
+
+    date_cols : List[int, str] (for "multicol" mode)
+        List of columns to be combined to construct the datetime values
+
+    col_units : str (for "table" mode)
+        Time units for the columns
+
+    zero_idx : bool (for "table" mode)
```

Comment:
Does this mean you expect the data to *not* have column names? i.e. you'll work off the column number alone? Is it not more likely that the value will be given as column names? If column names are given (preferred), this argument isn't needed.

Reply:
The example I have is (as shown in the docstring) with the column names as hrs:

| date        | 1     |   2   | 3      | .. |23      | 24     |
|-------------|-------|-------|-------|---|---------|--------|
| 2001/1/1 | val1 |val2 | val3 | .. | val23 | val24 |

Reply:
Is `zero_idx` not a bit redundant, then? Or I'm just confused as to its purpose :P

Reply:
When `zero_idx` is `True` the numbering is `0...23`, and `1...24` when `False`; in other words, that's the actual hour value as opposed to `n`th hour.

Reply:
hmm, this can get quite confusing, since different people mean different things from their (e.g.) hour numbering. One could refer to the starting point of the data, the end point, or the mid point (e.g. '1' might mean all data averaged from 0-1, 1-2, or 0.5-1.5). It's not always apparent. Also, '24' in the data might refer to midnight of that day, not the 24th hour. That would mean that turning it into '23' would be wrong, it would actually need to be turned into '0'. I wonder how these oddities could be covered by this API? Ideally it would aim to conform on one meaning for the timestamp (e.g. '00:00' = average from '00:00' to '01:00'), but `zero_idx` doesn't provide a user sufficient flexibility to enforce that (e.g. 24->0). It's also a confusing argument name, based on my initial misunderstanding...

Reply:
Note that this becomes even more confusing in models that group several hours (e.g., have 8 time steps per day, where `00:00` is the time period from midnight until 3am. Is the associated value then the value at midnight, at 3am, or an average...?

Reply:
@brynpickering @danielhuppmann I think both of you are reading too much into the implementation, looking at it as an analyst/modeller.  The point of this API is to convert non-standard structures into standard ones, that's it; no interpretation, no semantics.  All of that is left to the modeller.  The point about 24 meaning midnight is valid, but I don't know if it is common for say a power company to publish their data this way. Couple of examples of sample input files to BSAM are:
- hourly solar power generation, where the columns are numbered `1..24`
- hourly electricity import prices between two countries; columns are numbered `1..24`

Essentially this particular function is doing a transformation from a 2x2 matrix -> 1-D array.  Whatever was the column naming scheme is being encoded into the index **as is**, where for slightly more adaptability I've included the `zero_idx` flag; essentially a flag to undo my shift by one.